bug(comp): fix ambiguous truthiness in Compensation.apply

gegnew opened the pull request:
- bug(comp): fix ambiguous truthiness in Compensation.apply

gingo00 reviewed (approved):
Fix works for me.  My minor comments are not directly related to it, but more to the code of overall compensation.apply method.

The error message ("No channels from this file match those in the compensation.") could be improved, because that it should be more something around the lines of "The compensation matrix is not compatible with this file : All channels from the compensation matrix are not found in this file."

gegnew commented on cellengine/resources/compensation.py:
```diff
@@ -113,15 +113,18 @@ def apply(self, file: "FcsFile", inplace: bool = False, **kwargs):
+        )  # type: ignore
 
         # spill -> comp by inverting
         inverted = numpy.linalg.inv(self.dataframe)
```
This is a good idea.

Just a fun aside, this line is one of the very first pieces of code from the toolkit, from Zach's first investigations.

zbjornson replied:
I don't know what linalg library numpy uses, but MKL is quite fast at it for typical comp matrix sizes: 0.0047 ms for a 28 x 28 matrix.

gegnew replied:
It appears that `numpy` uses OpenBLAS if you install from Pypi, but MKL if you install via conda. But MKL is 700MB and OpenBLAS is 30MB. Probably not a huge need to use MKL directly here?

https://numpy.org/install/

zbjornson replied:
No need for MKL, no, was just posting a point of comparison.

gegnew commented on cellengine/resources/compensation.py:
```diff
@@ -113,15 +113,18 @@ def apply(self, file: "FcsFile", inplace: bool = False, **kwargs):
+        data: DataFrame = file.get_events(
+            **kwargs, inplace=True, destination=None
+        )  # type: ignore
 
```
I think that this should take `events` as an arg instead of `fcs_file`, and return the comped events. Does that make more sense?

gingo00 replied:
I guess it could.  (  applyToEvents vs applyToFcsfile).  Its the "inplace" that threw me off because if it updates "inplace", then I expected it to start from the existing event df.

zbjornson replied:
Maybe do something like:
If `file._events` use those AND ideally assert that kwargs here are the same as the kwargs that were used when calling `file.get_events`. Requires storing the kwargs for the latter, which could actually be useful.
Else call `file.get_events()`

gegnew replied:
Are the changes in 0332b79 and 9f7446f what you intended? These counteract the changes I made in ffff4f6, where I changed `.apply()` to accept an events dataframe; I'm not sure whether I like this better.

zbjornson replied:
Those commits don't exist anymore, which ones do you mean?

zbjornson commented on cellengine/resources/compensation.py:
```diff
@@ -100,36 +101,50 @@ def dataframe_as_html(self):
                 the target FcsFile
+
+        Returns:
+            DataFrame: The compensated events.
```
Duplicate `Returns:` now.

zbjornson commented on cellengine/resources/compensation.py:
```diff
@@ -100,36 +101,50 @@ def dataframe_as_html(self):
         """
-        data = file.get_events(**kwargs, inplace=True, destination=None)
+
+        if kwargs.items() == file._events_kwargs.items():
```
Let's document this behavior in the `kwargs` line.

> If the file's events have already been retrieved with the same kwargs provided here, those stored events will be used.

zbjornson commented on cellengine/resources/compensation.py:
```diff
@@ -100,36 +101,50 @@ def dataframe_as_html(self):
-            comped = data[self.channels]
+        cols = data.columns
+        ix = list(
+            filter(
+                None,
+                [channel if channel in cols else None for channel in self.channels],
+            )
+        )
+        if any(ix):
+            comped = data[ix]
             comped = comped.dot(inverted)  # type: ignore
             comped.columns = self.channels
             data.update(comped)
         else:
             raise IndexError(
                 "No channels from this file match those in the compensation."
             )
-
-        if inplace:
-            file._events = data
-        else:
-            return data
+        return data
```
Think this should be reverted, right?

gegnew replied:
It should be; I left it in so you can compare the two implementations. I'll clean it up if we prefer checking kwargs

zbjornson commented on tests/unit/resources/test_compensation.py:
```diff
@@ -1,18 +1,34 @@
+    comp.update(
+        {
+            "experimentId": EXP_ID,
+            # "channels": fcs_file.channels,
+            # "spillMatrix": identity(len(fcs_file.channels), int).flatten().tolist()
```
WIP here?

zbjornson commented on tests/unit/resources/test_compensation.py:
```diff
@@ -98,3 +108,143 @@ def test_create_from_spill_string(spillstring):
+
+
+@responses.activate
+def test_apply_scale_errors_for_nonmatching_channels(
```
```suggestion
def test_apply_comp_errors_for_nonmatching_channels(
```

zbjornson commented on tests/unit/resources/test_compensation.py:
```diff
@@ -98,3 +108,143 @@ def test_create_from_spill_string(spillstring):
+
+
+@responses.activate
+def test_apply_compensation_to_fcs_file(compensation, fcs_file):
```
Silly question: where do `compensation` and `fcs_file` come from? I assume it's the return of `def compensation(ENDPOINT_BASE, client, fcs_file, compensations)`, but then where does `compensations` come from?

zbjornson replied:
This test can probably be removed. I don't think there are any branches in the code that this covers vs. `test_apply_scale_compensates_values`.

zbjornson commented on tests/unit/resources/test_compensation.py:
```diff
@@ -98,3 +108,143 @@ def test_create_from_spill_string(spillstring):
+    # When: a Compensation is applied
+    results = compensation.apply(fcs_file)
+
+    # Then: events should be scaled
```
```suggestion
    # Then: events should be compensated
```

zbjornson commented on tests/unit/resources/test_compensation.py:
```diff
@@ -98,3 +108,143 @@ def test_create_from_spill_string(spillstring):
+    # When: a Compensation is applied
+    results = compensation.apply(fcs_file, testKwarg="foo")
+
+    # Then: events should be scaled
```
```suggestion
    # Then: events should be compensated
```

zbjornson commented on tests/unit/resources/test_compensation.py:
```diff
@@ -98,3 +108,143 @@ def test_create_from_spill_string(spillstring):
+    # When: a Compensation is applied
+    results = compensation.apply(fcs_file, inplace=False)
+
+    # Then: events should be scaled
```
```suggestion
    # Then: events should be compensated
```

zbjornson commented on tests/unit/resources/test_compensation.py:
```diff
@@ -98,3 +108,143 @@ def test_create_from_spill_string(spillstring):
+    results = compensation.apply(fcs_file, inplace=False)
+
+    # Then: events should be scaled
+    assert all(results["FSC-H"] != test_col)
```
This test should be a bit more rigorous.

I'd upload the Novocyte file to CellEngine, then go to Export Populations and pick "TSV" format, compensated, file-internal comp. You'll get a TSV file that you can use as the approximate truth values (likely within 1e-5). Then in this test you can create the comp from the file's spill string and assert on the values of the first row of the dataframe.

zbjornson commented on tests/unit/resources/test_compensation.py:
```diff
@@ -98,3 +108,143 @@ def test_create_from_spill_string(spillstring):
+
+
+@responses.activate
+def test_apply_scale_compensates_values(client, ENDPOINT_BASE, compensation, fcs_file):
```
```suggestion
def test_apply_comp_compensates_values(client, ENDPOINT_BASE, compensation, fcs_file):
```

zbjornson commented on cellengine/resources/compensation.py:
```diff
@@ -30,6 +30,7 @@ class Compensation(DataClassMixin):
             decoder=lambda x: numpy.array(x),
         )
     )
+    _inverted: Any = field(default=None)
```
Minor, but this can probably be `Union[None, Dataframe]` or `Optional[Dataframe]`?

zbjornson commented on tests/unit/resources/test_compensation.py:
```diff
@@ -98,3 +108,143 @@ def test_create_from_spill_string(spillstring):
+
+
+@responses.activate
+def test_apply_scale_with_subset_of_values(
```
If you improve the assertion in `test_apply_scale_compensates_values`, then you can delete this test also. The file-internal comp in the Novocyte file has 18 of the file's 24 channels.

zbjornson reviewed (changes requested):
Alright, last round, just test comments.

zbjornson commented on cellengine/resources/fcs_file.py:
```diff
@@ -14,16 +13,18 @@
         metadata=config(field_name="_id"), default=ReadOnly()
     )  # type: ignore
     compensation: Optional[int] = None
+    is_control: Optional[str] = None
```
Sorry, unrelated but noticed it because the line moved. `Optional[bool]`.

zbjornson commented on tests/unit/resources/test_compensation.py:
```diff
@@ -98,3 +108,100 @@ def test_create_from_spill_string(spillstring):
+
+
+@responses.activate
+def test_apply_compensation_to_fcs_file_with_no_events(
```
I think this is the test I suggested removing. The test title here and comment on 134 don't match the actual test behavior. Should be adequately covered by the new test at the end.

130-151 ✂️

zbjornson commented on tests/unit/resources/test_compensation.py:
```diff
@@ -98,3 +108,100 @@ def test_create_from_spill_string(spillstring):
+        [
+            isclose(a, b)
+            for a, b in zip(
+                results.head(1).values.tolist()[0],
+                acea_events_compensated.head(1).values.tolist()[0],
```
`head(1)` is selecting the first row, yeah? What's `[0]` selecting? (Not just the first column, right?)

If you have the entire 30 MB FCS file checked in, you might as well check a few more rows if it's just a matter of changing `head(1)` to `head(5)`.

zbjornson commented on tests/fixtures/compensated_events.py:
```diff
@@ -0,0 +1,35 @@
+@pytest.fixture(scope="function")
+def acea_events_compensated(acea_events):
+    """File-internal compensated events from 'Acea - Novocyte.fcs'"""
+    file = read_csv("tests/data/Acea_compensated.zip", sep="\t")
```
Not a priority right now, but if anyone complains about package size, it's a waste to check in a 30 MB file to only use the first row. Could just use `head -n 5` on the file or hard-code the values into a .py file here.

zbjornson reviewed (changes requested):
Almost there, sorry to drag this out.

gegnew commented:
I think I've addressed the previous three comments. Answers to the questions there, in order:

You suggested removing `test_apply_compensation_to_fcs_file`, but at any rate this is also redundant.

`head(1)` selected the first row. `.values.tolist()` returned a nested list `[['values']]` and then `[0]` was just getting the actual list of values. Simplified in the latest iteration, as both `results` and `acea_events_compensated` uniformly return `float` as the `dtype`

zbjornson commented on cellengine/resources/compensation.py:
```diff
@@ -100,30 +101,38 @@ def dataframe_as_html(self):
+            inverted = self._inverted
+        else:
+            inverted = numpy.linalg.inv(self.dataframe)
+            self._inverted = inverted
```
Gahh one last thing. When `self.dataframe` changes, `self._inverted` needs to be invalidated.

zbjornson reviewed (changes requested):
Found one bug, sorry.

zbjornson commented on tests/unit/resources/test_compensation.py:
```diff
@@ -98,3 +108,72 @@ def test_create_from_spill_string(spillstring):
+    results = acea_compensation.apply(acea_fcs_file, inplace=False)
+
+    # Then: events should be compensated correctly
+    assert all(results.head(5) == acea_events_compensated.head(5))
```
This test doesn't seem to be asserting correctly. It looked suspicious that these were digit-for-digit identical between CellEngine and Python. I changed a few numbers in acea_compensated.json and the test still passes.
![image](https://user-images.githubusercontent.com/469365/144933566-0ccc7faf-0e4f-4935-b479-0500879e64e8.png)

gingo00 replied:
Possibly a good use case for [`pandas.testing.assert_frame_equal`](https://pandas.pydata.org/docs/reference/api/pandas.testing.assert_frame_equal.html) , which supports approximate (with tolerances) matching.

zbjornson commented:
There we go. Yes, I changed the highlighted values manually.

There's generally no reason to use float64 for FCS data because *all* vendors use 32-bit datatypes for modern data (ints for old data). Looks like we can tell pandas to do that; something to look at later.

zbjornson reviewed (approved):
Looks good.

Rebased and squashed.
